File: TSA_Rail_System_TunnelProject/Wdw_Sust.OpcDa.Core/OpcTagGeneratorBase.vb
Imports System.Reflection
Imports TSA_Rail_System.Attributes
Imports TSA_Rail_System.Interfaces
Imports Wdw_Sust.Core.Attributes
Imports Wdw_Sust.OpcDa.Core.Interfaces

Public MustInherit Class OpcTagGeneratorBase
    Implements IOpcTagGenerator

    ''' <summary>
    ''' Used to cache tag definitions so that it is only generated once. This is a shared field
    ''' </summary>
    Private Shared ReadOnly _tagDefinitionCache As New Dictionary(Of Type, List(Of String))()

    ''' <summary>
    ''' Logic that generates the opc tags.
    ''' </summary>
    ''' <param name="obj">The object.</param>
    ''' <returns>System.String().</returns>
    ''' <exception cref="System.Exception">PLC name is not defined!!</exception>
    Public Function GenerateOpcTags(obj As IOpcObject) As String() Implements IOpcTagGenerator.GenerateOpcTags
        If String.IsNullOrEmpty(obj.PlcName) Then
            Throw New Exception("PLC name is not defined!!")
        End If
        Dim filters As String() = obj.FilterProperties
        Dim names As New List(Of String)
        Dim s As String = String.Empty
        'If this is OpcElemnt class, no need to use cache
        If obj.ListOfOpcTags.Count > 0 Then
            For Each k As String In obj.ListOfOpcTags.Keys
                names.Add(k)
            Next
            Return names.ToArray
            'If this is an OpcElement
        ElseIf TypeOf obj Is OpcElement Then
            s = GenerateOpcElementTagName(obj)
            'TODO: this needs to be separated from GetPLCTagNames as it intends to return a list of plc tags
            obj.ListOfOpcTags.Add(s, 1)  ' OpcElement always has 1 index only
            names.Add(s)
            If OpcComm.PrintTagNameWhenDebug Then Console.WriteLine(String.Format("{0} = index {1}", s, 1))
            Return names.ToArray
            'If list of tags has already in the cache
        ElseIf _tagDefinitionCache.ContainsKey(obj.GetType) Then
            names = _tagDefinitionCache(obj.GetType)
            'If the list of the tags for the class has not been defined yet
        Else
            Dim publicProperties() As PropertyInfo
            Dim myType As Type = obj.GetType()
            publicProperties = myType.GetProperties()
            Dim i As Integer
            For i = 0 To publicProperties.Length - 1
                If Attribute.IsDefined(publicProperties(i), GetType(IsOpcArrayAttribute)) Then
                    Dim a As IsOpcArrayAttribute = Attribute.GetCustomAttribute(publicProperties(i), GetType(IsOpcArrayAttribute))
                    For j As Integer = a.ArrayStartIndex To (a.ArraySize - 1 + a.ArrayStartIndex)
                        TagNameBreakDown(filters, names, publicProperties(i), , j)
                    Next
                Else
                    TagNameBreakDown(filters, names, publicProperties(i))
                End If
            Next
            'Added 6/1/2017 to handle the control class tag generations
            '
            If names.Count = 0 Then
                Dim tempTag As String = GenerateOpcElementTagName(obj)
                names.Add(tempTag)
                obj.ListOfOpcTags.Add(tempTag, 1)
                If OpcComm.PrintTagNameWhenDebug Then Console.WriteLine(String.Format("{0} = index {1}", s, 1))
                Return names.ToArray
            End If

            names.Sort()
            'Add this to the cache
            _tagDefinitionCache.Add(obj.GetType, names)
        End If
        Dim index As Integer = 1
        Dim tempList As New List(Of String)
        For Each s In names
            s = GenerateOpcTagName(obj.PlcName, obj.TagName, s)
            tempList.Add(s)
            obj.ListOfOpcTags.Add(s, index)
            If OpcComm.PrintTagNameWhenDebug Then 'TODO: How to handle Allen Bradley comm hard coded.
                Console.WriteLine(String.Format("{0} = index {1}", s, index))
            End If
            index += 1
        Next
        Return tempList.ToArray
        tempList = Nothing
    End Function

    Protected MustOverride Function GenerateOpcElementTagName(obj As IOpcObject) As String

    Protected MustOverride Function GenerateOpcTagName(plcTopicName As String, tagName As String, subTagName As String) As String

    ''' <summary>
    ''' Used to filter out the properties which are not part of the Opc item structure
    ''' </summary>
    ''' <param name="propertyName">Name of the property</param>
    ''' <returns>Boolean to indicate whether the property is pre-defined</returns>
    ''' <remarks></remarks>
    Private Function IsPreDefinedProperties(propertyName As String, filters As String()) As Boolean
        For Each s As String In filters
            If s = propertyName Then
                Return True
            End If
        Next
        Return False
    End Function

    Private Function IsPropertyOpcRelated(publicProperty As PropertyInfo, filters As String()) As Boolean
        'Check if the property of the object falls in the scope of WDW_SUST.dll
        If Not publicProperty.Module.ScopeName.ToLower.StartsWith("wdw_sust") Then Return False ' And Not publicProperty.ReflectedType.BaseType.FullName = "WDW_SUST.Textile.LINT.Opc.Core.OpcObjectBase" Then Return False
        'Check if the property is non-connectable Opc item
        If Attribute.IsDefined(publicProperty, GetType(NotConnectableOpcItemAttribute)) Then
            If Not Attribute.IsDefined(publicProperty, GetType(ConnectableOpcItemAttribute)) Then Return False
        End If
        'Check if the property is predefined to be not connected
        If IsPreDefinedProperties(publicProperty.Name, filters) Then Return False
        Return True
    End Function

    ''' <summary>
    ''' Subroutine to break down the tag name to the element level
    ''' </summary>
    Private Sub TagNameBreakDown(filters As String(), ByRef names As List(Of String), publicProperty As PropertyInfo, Optional subTagName As String = "", Optional index As Integer = -1)
        If IsPropertyOpcRelated(publicProperty, filters) Then
            If (publicProperty.PropertyType.IsClass OrElse publicProperty.PropertyType.IsInterface) AndAlso
                publicProperty.PropertyType.FullName <> "System.String" Then
                If subTagName Is String.Empty Then
                    If index > -1 Then
                        subTagName = String.Format("{0}[{1}]", publicProperty.Name, index)
                    Else
                        subTagName = String.Format("{0}", publicProperty.Name)
                    End If
                Else
                    subTagName = String.Format("{0}.{1}", subTagName, publicProperty.Name)
                End If
                Dim mySubType As Type = publicProperty.PropertyType
                Dim subPublicProperties() As PropertyInfo
                subPublicProperties = mySubType.GetProperties

                For i As Integer = 0 To subPublicProperties.Length - 1
                    If Attribute.IsDefined(subPublicProperties(i), GetType(IsOpcArrayAttribute)) Then
                        Dim a As IsOpcArrayAttribute = Attribute.GetCustomAttribute(subPublicProperties(i), GetType(IsOpcArrayAttribute))
                        For j As Integer = a.ArrayStartIndex To (a.ArraySize - 1 + a.ArrayStartIndex)
                            TagNameBreakDown(filters, names, subPublicProperties(i), subTagName, j)
                        Next
                    Else
                        TagNameBreakDown(filters, names, subPublicProperties(i), subTagName)
                    End If
                Next
            ElseIf subTagName Is String.Empty Then
                If index > -1 Then
                    subTagName = String.Format("{0}[{1}]", publicProperty.Name, index)
                Else
                    subTagName = String.Format("{0}", publicProperty.Name)
                End If
                names.Add(subTagName)

            Else
                If index > -1 Then
                    subTagName = String.Format("{0}.{1}[{2}]", subTagName, publicProperty.Name, index)
                Else
                    subTagName = String.Format("{0}.{1}", subTagName, publicProperty.Name)
                End If
                names.Add(subTagName)
            End If

        End If
    End Sub

End Class

File: Wdw_Sust.OpcDa.Core.Plc/AllenBradley/OpcTagGeneratorBase.cs
using System.Reflection;
//using TSA_Rail_System.Attributes;
//using TSA_Rail_System.Interfaces;
//using Wdw_Sust.Core.Attributes;
//using Wdw_Sust.OpcDa.Core.Interfaces;

namespace Wdw_Sust.OpcDa.Core.Plc.AllenBradley
{
    /// <summary>
    /// Abstract base class that implements IOpcTagGenerator.
    /// Provides core logic for generating OPC tags from IOpcObject instances.
    /// </summary>
    public abstract class OpcTagGeneratorBase : IOpcTagGenerator
    {
        /// <summary>
        /// Used to cache tag definitions so that they are generated only once.
        /// </summary>
        private static readonly Dictionary<Type, List<string>> _tagDefinitionCache = new();

        /// <summary>
        /// Generates OPC tags for the provided object.
        /// </summary>
        /// <param name="obj">The OPC object.</param>
        /// <returns>Array of OPC tag names.</returns>
        /// <exception cref="Exception">Thrown when PlcName is not defined.</exception>
        public string[] GenerateOpcTags(IOpcObject obj)
        {
            if (string.IsNullOrEmpty(obj.PlcName))
                throw new Exception("PLC name is not defined!!");

            var filters = obj.FilterProperties;
            var names = new List<string>();

            // If object already has tags, return them
            if (obj.ListOfOpcTags.Count > 0)
            {
                foreach (var k in obj.ListOfOpcTags.Keys)
                    names.Add(k);

                return names.ToArray();
            }

            // If this is a primitive OPC element
            if (obj is OpcElement)
            {
                var tag = GenerateOpcElementTagName(obj);
                obj.ListOfOpcTags.Add(tag, 1);
                names.Add(tag);

                if (OpcComm.PrintTagNameWhenDebug)
                    Console.WriteLine($"{tag} = index 1");

                return names.ToArray();
            }

            // If type is already cached
            if (_tagDefinitionCache.ContainsKey(obj.GetType()))
            {
                names = _tagDefinitionCache[obj.GetType()];
            }
            else
            {
                var myType = obj.GetType();
                var publicProperties = myType.GetProperties();

                foreach (var property in publicProperties)
                {
                    if (Attribute.IsDefined(property, typeof(IsOpcArrayAttribute)))
                    {
                        var arrayAttr = (IsOpcArrayAttribute)Attribute.GetCustomAttribute(property, typeof(IsOpcArrayAttribute))!;
                        for (int j = arrayAttr.ArrayStartIndex; j <= arrayAttr.ArraySize - 1 + arrayAttr.ArrayStartIndex; j++)
                        {
                            TagNameBreakDown(filters, names, property, string.Empty, j);
                        }
                    }
                    else
                    {
                        TagNameBreakDown(filters, names, property);
                    }
                }

                // Handle objects with no tags
                if (names.Count == 0)
                {
                    var tempTag = GenerateOpcElementTagName(obj);
                    names.Add(tempTag);
                    obj.ListOfOpcTags.Add(tempTag, 1);

                    if (OpcComm.PrintTagNameWhenDebug)
                        Console.WriteLine($"{tempTag} = index 1");

                    return names.ToArray();
                }

                names.Sort();
                _tagDefinitionCache.Add(obj.GetType(), names);
            }

            var index = 1;
            var finalTags = new List<string>();
            foreach (var rawName in names)
            {
                var tagName = GenerateOpcTagName(obj.PlcName, obj.TagName, rawName);
                finalTags.Add(tagName);
                obj.ListOfOpcTags.Add(tagName, index);

                if (OpcComm.PrintTagNameWhenDebug)
                    Console.WriteLine($"{tagName} = index {index}");

                index++;
            }

            return finalTags.ToArray();
        }

        /// <summary>
        /// Generates the OPC element tag name for a primitive object.
        /// </summary>
        protected abstract string GenerateOpcElementTagName(IOpcObject obj);

        /// <summary>
        /// Combines PLC topic name, tag name, and sub-tag name into a full OPC tag.
        /// </summary>
        protected abstract string GenerateOpcTagName(string plcTopicName, string tagName, string subTagName);

        /// <summary>
        /// Handles formatting of OPC array tags.
        /// </summary>
        public abstract string HandleOpcArrayTagName(string tagName, string propertyName, int index);

        /// <summary>
        /// Parameterless version of GenerateOpcTagName required by interface.
        /// </summary>
        public string[] GenerateOpcTagName()
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Checks if the property name is predefined to be excluded.
        /// </summary>
        private static bool IsPreDefinedProperty(string propertyName, string[] filters)
        {
            foreach (var s in filters)
            {
                if (s == propertyName) return true;
            }
            return false;
        }

        /// <summary>
        /// Determines if a property is related to OPC tag generation.
        /// </summary>
        private bool IsPropertyOpcRelated(PropertyInfo publicProperty, string[] filters)
        {
            if (!publicProperty.Module.ScopeName.ToLower().StartsWith("wdw_sust"))
                return false;

            if (Attribute.IsDefined(publicProperty, typeof(NotConnectableOPCItemAttribute)) &&
                !Attribute.IsDefined(publicProperty, typeof(ConnectableOpcItemAttribute)))
                return false;

            if (IsPreDefinedProperty(publicProperty.Name, filters))
                return false;

            return true;
        }

        /// <summary>
        /// Breaks down property structure into tag names recursively.
        /// </summary>
        private void TagNameBreakDown(string[] filters, List<string> names, PropertyInfo publicProperty, string subTagName = "", int index = -1)
        {
            if (!IsPropertyOpcRelated(publicProperty, filters))
                return;

            if ((publicProperty.PropertyType.IsClass || publicProperty.PropertyType.IsInterface) &&
                publicProperty.PropertyType.FullName != "System.String")
            {
                if (string.IsNullOrEmpty(subTagName))
                {
                    if (index > -1)
                        subTagName = $"{publicProperty.Name}[{index}]";
                    else
                        subTagName = publicProperty.Name;
                }
                else
                {
                    subTagName = $"{subTagName}.{publicProperty.Name}";
                }
                subTagName = HandleOpcArrayTagName(subTagName, publicProperty.Name, index);

                var subProperties = publicProperty.PropertyType.GetProperties();
                foreach (var subProp in subProperties)
                {
                    if (Attribute.IsDefined(subProp, typeof(IsOpcArrayAttribute)))
                    {
                        var arrayAttr = (IsOpcArrayAttribute)Attribute.GetCustomAttribute(subProp, typeof(IsOpcArrayAttribute))!;
                        for (int j = arrayAttr.ArrayStartIndex; j <= arrayAttr.ArraySize - 1 + arrayAttr.ArrayStartIndex; j++)
                        {
                            TagNameBreakDown(filters, names, subProp, subTagName, j);
                        }
                    }
                    else
                    {
                        TagNameBreakDown(filters, names, subProp, subTagName);
                    }
                }
            }
            else if (string.IsNullOrEmpty(subTagName))
            {
                if (index > -1)
                    subTagName = $"{publicProperty.Name}[{index}]";
                else
                    subTagName = publicProperty.Name;

                names.Add(subTagName);
            }
            else
            {
                if (index > -1)
                    subTagName = $"{subTagName}.{publicProperty.Name}[{index}]";
                else
                    subTagName = $"{subTagName}.{publicProperty.Name}";

                names.Add(subTagName);
            }
        }
    }
}
